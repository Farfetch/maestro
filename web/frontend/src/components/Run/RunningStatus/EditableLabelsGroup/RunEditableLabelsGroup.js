import { PlusOutlined } from "@ant-design/icons";
import { Button, Input, message, Tag, Tooltip } from "antd";
import PropTypes from "prop-types";
import { useEffect, useRef, useState } from "react";

import { updateRun } from "../../../../lib/api/endpoints/run";

const RunEditableLabelsGroup = ({ runId, defaultValue = [] }) => {
  const [labels, setLabels] = useState(defaultValue);
  const [inputVisible, setInputVisible] = useState(false);
  const [inputValue, setInputValue] = useState("");
  const input = useRef(null);

  const updateRunLabels = async (labelsToUpdate) => {
    await updateRun(runId, {
      labels: labelsToUpdate
    });
    message.success({ content: "Saved!" });
  };
  const handleClose = async (removedTag) => {
    const newLabels = labels.filter((tag) => tag !== removedTag);

    setLabels(newLabels);
    await updateRunLabels(newLabels);
  };

  const handleInputConfirm = async () => {
    if (inputValue && labels.indexOf(inputValue) === -1) {
      const newLabels = [...labels, inputValue];
      setLabels(newLabels);
      await updateRunLabels(newLabels);
    }

    setInputVisible(false);
    setInputValue("");
  };

  useEffect(() => {
    if (inputVisible) {
      input.current.focus();
    }
  }, [inputVisible]);

  useEffect(() => {}, [labels]);

  return (
    <>
      {labels.map((tag) => {
        const isLongTag = tag.length > 20;

        const tagElem = (
          <Tag key={tag} closable={true} onClose={() => handleClose(tag)}>
            <span>{isLongTag ? `${tag.slice(0, 20)}...` : tag}</span>
          </Tag>
        );
        return isLongTag ? (
          <Tooltip title={tag} key={tag}>
            {tagElem}
          </Tooltip>
        ) : (
          tagElem
        );
      })}
      {inputVisible && (
        <Input
          ref={input}
          type="text"
          size="small"
          style={{ width: "160px" }}
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
          onBlur={handleInputConfirm}
          onPressEnter={handleInputConfirm}
        />
      )}
      {!inputVisible && (
        <Button
          type="dashed"
          size="small"
          icon={<PlusOutlined />}
          onClick={() => setInputVisible(true)}
        >
          New label
        </Button>
      )}
    </>
  );
};

RunEditableLabelsGroup.propTypes = {
  runId: PropTypes.string.isRequired,
  defaultValue: PropTypes.arrayOf(PropTypes.string)
};

export default RunEditableLabelsGroup;
