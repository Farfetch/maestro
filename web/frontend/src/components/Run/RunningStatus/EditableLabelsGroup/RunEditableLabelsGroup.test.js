import { render } from "@testing-library/react";
import React from "react";

import RunEditableLabelsGroup from "./RunEditableLabelsGroup";

describe("components/Run/RunningStatus/EditableLabelsGroup", () => {
  const runId = "test";
  test(`should render RunEditableLabelsGroup component`, async () => {
    const rendered = render(
      <RunEditableLabelsGroup runId={runId} defaultValue={[]} />
    );

    const badgeComponent = rendered.container;
    expect(badgeComponent.outerHTML).toMatchSnapshot();
  });

  test(`should render RunEditableLabelsGroup component with default value`, async () => {
    const rendered = render(
      <RunEditableLabelsGroup
        runId={runId}
        defaultValue={["label1", "label2"]}
      />
    );

    const badgeComponent = rendered.container;
    expect(badgeComponent.outerHTML).toMatchSnapshot();
  });
});
