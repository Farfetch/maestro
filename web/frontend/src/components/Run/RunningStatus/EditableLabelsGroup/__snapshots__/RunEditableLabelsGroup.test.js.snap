// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Run/RunningStatus/EditableLabelsGroup should render RunEditableLabelsGroup component 1`] = `"<div><button type=\\"button\\" class=\\"ant-btn ant-btn-dashed ant-btn-sm\\"><span role=\\"img\\" aria-label=\\"plus\\" class=\\"anticon anticon-plus\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"plus\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><defs><style></style></defs><path d=\\"M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z\\"></path><path d=\\"M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z\\"></path></svg></span><span>New label</span></button></div>"`;

exports[`components/Run/RunningStatus/EditableLabelsGroup should render RunEditableLabelsGroup component with default value 1`] = `"<div><span class=\\"ant-tag\\"><span>label1</span><span role=\\"img\\" aria-label=\\"close\\" tabindex=\\"-1\\" class=\\"anticon anticon-close ant-tag-close-icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></span><span class=\\"ant-tag\\"><span>label2</span><span role=\\"img\\" aria-label=\\"close\\" tabindex=\\"-1\\" class=\\"anticon anticon-close ant-tag-close-icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></span><button type=\\"button\\" class=\\"ant-btn ant-btn-dashed ant-btn-sm\\"><span role=\\"img\\" aria-label=\\"plus\\" class=\\"anticon anticon-plus\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"plus\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><defs><style></style></defs><path d=\\"M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z\\"></path><path d=\\"M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z\\"></path></svg></span><span>New label</span></button></div>"`;
